lib/vrl/DESIGN.md

```diff
@@ -0,0 +1,219 @@
+
+Its purpose is to hit a sweet spot between Turing complete languages such as
+Lua, and static transforms such as Vector's `rename_fields`. It should be
+flexible enough to cover most remapping use-cases, without requiring operators
+to write complex scripts that are difficult to reason about and incur
+a significant performance penalty.
+
+See the [introduction blog post][blog] for more details on the **why**.
+
+[blog]: https://vector.dev/blog/vector-remap-language/#preamble
+
+## Target Audience
+
+VRL has a specialized purpose of solving observability data remapping needs.
+Because of this purpose, the language is mostly used by people who manage
+infrastructure at their organisations. The role of this group is usually
+referred to as "operator", "devops" (Development and Operations) or "SRE" (Site
+Reliability Engineer).
+
+One common generalization of this group is that they are focused on maintaining
+infrastructure within an organization, and often write and maintain their own
+software to achieve this goal. They are adept enough at programming to achieve
+their goals, but have no need or desire to be as skilled in programming as
+dedicated software engineers, because their time is best spent elsewhere.
+
+As with everything, there are exceptions to the rule, and many people in this
+group _are_ highly skilled software engineers, but VRL must capture the largest
+possible segment of this group, and should therefor be limited in complexity.
+
+Therefor, when extending the feature set of VRL, design **the feature for the
+intended target audience**, which will likely mean choosing different trade-offs
+than you'd make if you were to design the feature for your personal needs.
+
+## Language Limits
+
+There are a number of features that we've so far rejected to implement in the
+language. This might change in the future, but there should be a good reason to
+do so.
+
+- modules (see: [#5507][]
+- classes
+- user-defined functions
+- `goto` statements
+- network calls (see [#4517][])
+
+[#5507]: https://github.com/timberio/vector/issues/5507
+[#4517]: https://github.com/timberio/vector/issues/4517#issuecomment-754160338
+
+## Conventions
+
+### Functions
+
+#### Composition
+
+- Favor function composition over single-purpose functions.
+
+- If a problem needs to be solved in multiple steps, consider adding
+  single-purpose functions for each individual step.
+
+- If useability or readability is hurt by composition, favor single-purpose
+  functions.
+
+#### Naming
+
+- Function names are lower-cased (e.g. `round`).
+
+- Multi-name functions use underscore (`_`) as a separator (e.g.
+  `encode_key_value`).
+
+- Favor explicit verbose names over terse ones (e.g.
+  `parse_aws_cloudwatch_log_subscription_message` over `parse_aws_cwl_msg`).
+
+- Use `parse_*` for string decoding functions (e.g. `parse_json` and
+  `parse_grok`).
+
+- Use `encode_*` for string encoding functions (e.g. `encode_base64`).
```

Comment:
It should be noted that some decoding functions are named `decode_*` (`decode_base64` and `decode_percent`).

Ideally, we'd rectify this, but that's a breaking change. We can create issues for when a current implementation goes against the design principles so that we can do a sweep of breaking changes before we hit 1.0.

Alternatively, we accept this, document it, and keep things as is, even when we release 1.0.

Reply:
I think there is some distinction here between functions that parse a string into some other type (like `parse_json` or `parse_int`) and functions that turn a string into another string (`decode_base64` or `decode_percent`).

To me `parse_base64` reads pretty weird given no "parsing" is really happening, just translation. I think `parse` makes sense for only for structured data.

It does makes `encode` a little confusing since it is used as the opposite of both `decode` and `parse` when you might only expect it to be the opposite of `decode`. I might suggest we use `to_X` like `to_json` and `to_key_value` but this would overload `to_X` given it is also used for converting to primitives. We could go the Go/Java route of `unmarshal` and `marshal`?

> Ideally, we'd rectify this, but that's a breaking change. We can create issues for when a current implementation goes against the design principles so that we can do a sweep of breaking changes before we hit 1.0.

I think we could also just alias the functions to rename them to the desired scheme; dropping the deprecated aliases at 1.0.

Reply:
💭 `marshal`/`unmarshal` seem natural given my history in Go - but it is a larger change (aliasing seems reasonable)